### PR TITLE
feat: move to new config schema and location

### PR DESCRIPTION
**Pull request by cmoog**
No rush to merge this, but it will open up some nice opportunities when it lands.

This PR changes the way we store user config and credentials to support the growing scope of `coder-cli`. 

## Current State
```
~/.config/coder/session
~/.config/coder/url
```

## New State
`~/.coder/credentials.yaml`:
```yaml
url: https://master.cdr.dev
session: 1234321
```

## Future
`~/.coder/config.yaml`
```yaml
default:
  editor: vscode
  environment: dev1
```

### CODE REVIEW

File: internal/config/file.go [hunk 1]
```diff
@@ -19,8 +24,52 @@ func (f File) Read() (string, error) {
...
+	DashboardURL string `yaml:"url"`
+	SessionToken string `yaml:"session"`
+}
+
+type CoderConfig struct {
+	Version  string        `yaml:"version"`
+	Defaults CoderDefaults `yaml:"defaults"`
+}
+
+type CoderDefaults struct {
```
Comment (kylecarbs) [inline internal/config/file.go:63]: 👍

**Review by kylecarbs — APPROVED** [review]

**Review by fuskovic — APPROVED** [review]

**Comment by Emyrk** [comment]
I'm confused by the description. We are moving out of `~/.config`, but will return to it?
Is there a standard we are using like [`XDG_CONFIG_HOME`](https://wiki.archlinux.org/index.php/XDG_Base_Directory)?

**Comment by cmoog** [comment]
> I'm confused by the description. We are moving out of ~/.config, but will return to it?
Is there a standard we are using like XDG_CONFIG_HOME?

I'm proposing we move from `~/.config/coder` to `~/.coder`... this is more aligned with countless other tools. To name a few
- ~/.vscode
- ~/.yarn
- ~/.rustup
- ~/.docker
- ~/.cargo
- ~/.k9s
- ~/.mutagen
- ~/.vim

**Comment by Emyrk** [comment]
I assumed a lot of existing apps were using `$HOME/.<app>` because of legacy reasons and trying not to break their existing users.
I can live with `$HOME/.coder`, I've just heard the argument about cluttering `$HOME` with a ton of dotfiles.

File: internal/config/file.go [hunk 2]
```diff
@@ -19,8 +24,58 @@ func (f File) Read() (string, error) {
...
+// WriteYAML writes the file as yaml.
+func (f File) WriteYAML(in interface{}) error {
+	encoded, err := yaml.Marshal(in)
+	if err != nil {
+		return err
+	}
+	return f.Write(string(encoded))
+}
+
+// The following files configure and authenticate the coder-cli in ~/.coder/.
```
Comment (Emyrk) [inline internal/config/file.go:48]: nit: The `.` at the end of this sentence made me think all config files would be dotfiles.
I think linting complains if you don't end in a period, maybe `... coder-cli in '~/.coder/'.`

**Review by Emyrk — APPROVED** [review]

**Comment by cmoog** [comment]
Closing this out for now since it's gotten stale, may revisit later as the CLI matures.

**Pull request closed without merging**